File: NOVEMBER-2025/28-11-2025/2872.php
class Solution {

    /**
     * @param Integer $n
     * @param Integer[][] $edges
     * @param Integer[] $values
     * @param Integer $k
     * @return Integer
     */
    function maxKDivisibleComponents($n, $edges, $values, $k) {
        $g = array_fill(0, $n, []);
        foreach ($edges as $e) {
            $a = $e[0];
            $b = $e[1];
            $g[$a][] = $b;
            $g[$b][] = $a;
        }

        $ans = 0;

        $dfs = function($u, $p) use (&$dfs, &$g, &$values, $k, &$ans) {
            $sum = $values[$u];
            foreach ($g[$u] as $v) {
                if ($v === $p) continue;
                $sum += $dfs($v, $u);
            }
            if ($sum % $k === 0) {
                $ans++;
                return 0;
            }
            return $sum;
        };

        $dfs(0, -1);
        return $ans;
    }
}

File: NOVEMBER-2025/28-11-2025/2872.rb
# @param {Integer} n
# @param {Integer[][]} edges
# @param {Integer[]} values
# @param {Integer} k
# @return {Integer}
def max_k_divisible_components(n, edges, values, k)
  g = Array.new(n) { [] }

  edges.each do |a, b|
    g[a] << b
    g[b] << a
  end

  @ans = 0

  dfs = lambda do |u, p|
    sum = values[u]
    g[u].each do |v|
      next if v == p
      sum += dfs.call(v, u)
    end
    if sum % k == 0
      @ans += 1
      0
    else
      sum
    end
  end

  dfs.call(0, -1)
  @ans
end